Add a Nix flake

sellout opened the pull request:
This provides

- two overlays (one for emacsPackages, one for nixpkgs),
- a minimal example homeConfiguration for testing,
- a fully-featured package for agenix.el (byte compiles and runs tests (there are none yet)),
- a devShell including development tooling, and
- multiple checks (running linters).

sellout committed:
Add a Nix flake

This provides

- two overlays (one for emacsPackages, one for nixpkgs),
- a minimal example homeConfiguration for testing,
- a fully-featured package for agenix.el (byte compiles and runs tests (there
  are none yet)),
- a devShell including development tooling, and
- multiple checks (running linters).

sellout commented:
BTW, the values in .dir-locals.el needed to be set to satisfy the package linters. I tried to pick values that matched the style of the existing code, but am happy to change any and fix the lints in the code instead.

t4ccer committed:
Move inputs up

t4ccer committed:
Don't use URL literals

t4ccer committed:
Remove redundant let

t4ccer commented on flake.nix:
```diff
@@ -0,0 +1,199 @@
+
+      homeConfigurations.example = inputs.homeManager.lib.homeManagerConfiguration {
+        pkgs = import inputs.nixpkgs {
+          system = inputs.flake-utils.lib.system.aarch64-darwin;
```
Could we move it inside `perSystem` and test in the CI

sellout replied:
There is an issue for having garnix run these (garnix-io/issues#24), but `homeConfigurations` is like `nixosConfigurations` in that it doesn’t expect a system component in the attribute path. It usually makes sense – you have a configuration for a specific machine, so the system is fixed. But in this case, it would be nice. And the way that flake-utils inserts the `system` into the tree is a bit too magical to just move this there.

We could do something like
```nix
{
  homeConfigurations =
    builtins.listToAddrs
    (builtins.map (system: {
                    name = "${system}-example";
                    value = inputs.homeManager.lib.homeManagerConfiguration ...;
                  })
                  inputs.flake-utils.defaultSystems);
}
```
which I wouldn’t be opposed to.

t4ccer replied:
OK, there is no need to rush until garnix supports these

t4ccer commented on flake.nix:
```diff
@@ -0,0 +1,199 @@
+            inherit ELDEV_LOCAL src;
+
+            pname = "agenix";
+            version = "0.2";
```
Could we read `version` from the source file?

t4ccer replied:
Is there a better tool for that, than just `grep`ing?

sellout replied:
I’ve been doing like
```nix
let
  contents = builtins.readFile someRandomFile.mk;
  readField = regex: infix: let
    match =
      builtins.match ".*\\$\\(package\\)${infix}_${regex}=([^\n]*).*" contents;
  in
    if match == null
    then null
    else builtins.elemAt match 0;
in {}
```
which may or may not be a better tool than `grep` 😄

It’ll be simpler in our case, since it’s not trying to dynamically build a regex.

t4ccer commented on flake.nix:
```diff
@@ -0,0 +1,199 @@
+
+            instalCheckPhase = ''
+              runHook preInstallCheck
+              eldev --packaged test
```
Why isn't it also inside `checkPhase`?

t4ccer commented on flake.nix:
```diff
@@ -0,0 +1,199 @@
+        ];
+      };
+    }
+    // inputs.flake-utils.lib.eachDefaultSystem (system: let
```
Not a huge fan on `flake-utils`, but don't have time to move it to `flake.parts` now so that's fine

sellout replied:
Oh, I hadn’t heard of flake.parts … I’ll have to look into that. Thanks.

t4ccer commented on flake.nix:
```diff
@@ -0,0 +1,199 @@
+            (inputs.self.overlays.emacs final prev);
+        };
+
+        emacs = final: prev: efinal: eprev: {
+          agenix = inputs.self.packages.${final.system}.agenix-el;
+        };
```
Is there a point to expose it separately? What's the usecase someone would use `overlays.emacs` instead of `overlays.default`?

sellout replied:
I provide something like this for any package that’s encapsulated in some “withPackages” like thing. In this case, it would be if someone was not using `emacsPackagesFor` for some reason. `emacsPackagesFor` is very general, so It might not be useful. In the Nix/Haskell world, you need to add the Haskell package explicitly to each compiler you support, so if the `default` overlay misses one, or a user adds their own, then the `haskell` overlay needs to be used directly to make it available via `withPackages`.

t4ccer replied:
Ohh I see, indeed it can be useful for someone

t4ccer commented:
Also would be nice to add nix formatting check

sellout reviewed (commented):
Ah, I hadn’t seen that URL literals were going away!

Obviously the quoted version is fine either way, but I’ve wondered for other reasons what the `github:t4ccer/agenix.el` syntax is even called. (And, relevant to this, does it fall under “URL syntax” as described in [RFC 45](https://github.com/NixOS/rfcs/blob/master/rfcs/0045-deprecate-url-syntax.md)?). In [the `nix flake` docs](https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-flake.html#url-like-syntax) it’s described as “URL-like” and called a “flake reference”, but paths and URLs are also under the umbrella of flake references (and at least sometimes when they don’t even refer to a flake (`flake = false;`), so “input reference” might be a better name).

We should perhaps add `nixConfig.extra-experimental-features = ["no-url-literals"];` to the flake, so it gets caught automatically. And I guess this would also tell me whether `github:…` is considered a URL.

sellout reviewed (commented):
I don’t have a problem with this, and I know they’re typically at the top. I just want to mention my justification for having them at the bottom.

The inputs are mostly boilerplate, or they’re at least “deep” parts of the flake that you can generally not pay attention to. I try to organize the flake in order of user sophistication:

* description – casual curious person;
* overlays, homeConfigurations – consumer of the flake;
* apps, packages – more advanced flake users;
* devShells – people poking at the code; and
* checks, formatter, inputs, nixConfig – contributors to the project.

t4ccer commented:
Re: URL literals
[nil](https://github.com/oxalica/nil) told me that `"github:t4ccer/agenix.el"` (in quotes) is no longer an URL literal.

Re: Order of description/inputs/outputs
inputs not being at the top was really confusing for me, for a second I thought that the flake didn't declare inputs at all and was using implicit flake-registry-provided nixpkgs input only, which is a no-no. I know it's mostly a matter of taste, but that's what I'm used to.

sellout commented:
> [nil](https://github.com/oxalica/nil) told me that `"github:t4ccer/agenix.el"` (in quotes) is no longer an URL literal.

👍🏼 (I hadn’t seen nil, either – I’ve been using rnix – do you have a preference?)

> Re: Order of description/inputs/outputs inputs not being at the top was really confusing for me, for a second I thought that the flake didn't declare inputs at all and was using implicit flake-registry-provided nixpkgs input only, which is a no-no. I know it's mostly a matter of taste, but that's what I'm used to.

Ah, that’s a good point. Yeah, I never rely on the registry, but I can see how the unconventional `inputs` location can make one nervous 😄

t4ccer commented:
rnix doesn't seem to be actively maintained for a couple of months already, so I just use nil

t4ccer committed:
Auto-detect version from `agenix.el` file

sellout committed:
Add a nix-fmt check

sellout committed:
Switch the Nix language server from rnix to nil

sellout committed:
Remove a redundant linter

`elisp` runs multiple linters, including checkdoc, which is what Eldev’s `doc`
runs.

sellout committed:
Update the flake

t4ccer committed:
Move to `flake.part`

It's just my personal preference
